exchanges/kraken/kraken_websocket.go

```diff
@@ -0,0 +1,676 @@
+			highestLastUpdate = bidUpdateTime
+		}
+	}
+
+	ob.LastUpdated = highestLastUpdate
+	err := k.Websocket.Orderbook.LoadSnapshot(&ob, k.GetName(), true)
+	if err != nil {
+		log.Error(err)
+	}
+
+	k.Websocket.DataHandler <- exchange.WebsocketOrderbookUpdate{
+		Exchange: k.GetName(),
+		Asset:    krakenWsAssetType,
+		Pair:     channelData.Pair,
+	}
+}
+
+// wsProcessOrderBookUpdate updates an orderbook entry for a given currency pair
+func (k *Kraken) wsProcessOrderBookUpdate(channelData WebsocketChannelData, obData map[string]interface{}) {
+	ob, err := k.GetOrderbookEx(channelData.Pair, krakenWsAssetType)
+	if err != nil {
+		k.Websocket.DataHandler <- err
+	}
+	// Kraken ob data is timestamped per price, GCT orderbook data is timestamped per entry
+	// Using the highest last update time, we can attempt to respect both within a reasonable degree
+	var highestLastUpdate time.Time
+	// Ask data is not always sent
+	if _, ok := obData["a"]; ok {
+		askData := obData["a"].([]interface{})
+		for i := range askData {
+			asks := askData[i].([]interface{})
+			price, _ := strconv.ParseFloat(asks[0].(string), 64)
+			amount, _ := strconv.ParseFloat(asks[1].(string), 64)
+
+			if amount == 0 {
+				for j := 0; j < len(ob.Asks); j++ {
+					if ob.Asks[j].Price == price {
+						ob.Asks = append(ob.Asks[:j], ob.Asks[j+1:]...)
+						j--
+						continue
+					}
+				}
+			}
+			ob.Asks = append(ob.Asks, orderbook.Item{
+				Amount: amount,
+				Price:  price,
+			})
+
+			timeData, _ := strconv.ParseFloat(asks[2].(string), 64)
+			sec, dec := math.Modf(timeData)
+			askUpdatedTime := time.Unix(int64(sec), int64(dec*(1e9)))
+			if highestLastUpdate.Before(askUpdatedTime) {
+				highestLastUpdate = askUpdatedTime
+			}
+		}
+	}
+	// Bid data is not always sent
+	if _, ok := obData["b"]; ok {
+		bidData := obData["b"].([]interface{})
+		for i := range bidData {
+			bids := bidData[i].([]interface{})
+			price, _ := strconv.ParseFloat(bids[0].(string), 64)
+			amount, _ := strconv.ParseFloat(bids[1].(string), 64)
+
+			if amount == 0 {
+				for j := 0; j < len(ob.Bids); j++ {
+					if ob.Bids[j].Price == price {
+						ob.Bids = append(ob.Bids[:j], ob.Bids[j+1:]...)
+						j--
+						continue
+					}
+				}
+			}
+
+			ob.Bids = append(ob.Bids, orderbook.Item{
+				Amount: amount,
+				Price:  price,
+			})
+
+			timeData, _ := strconv.ParseFloat(bids[2].(string), 64)
+			sec, dec := math.Modf(timeData)
+			bidUpdatedTime := time.Unix(int64(sec), int64(dec*(1e9)))
+			if highestLastUpdate.Before(bidUpdatedTime) {
+				highestLastUpdate = bidUpdatedTime
+			}
+		}
+	}
+
+	if ob.LastUpdated.After(highestLastUpdate) {
+		log.Errorf("orderbook update out of order. Existing: %v, Attempted: %v", ob.LastUpdated, highestLastUpdate)
+		k.ResubscribeToChannel(channelData.Subscription, channelData.Pair)
+		return
+	}
+	ob.LastUpdated = highestLastUpdate
+	err = k.Websocket.Orderbook.LoadSnapshot(&ob, k.GetName(), true)
+	if err != nil {
+		log.Error(err)
+	}
```

Comment:
add an else so we don't send an update switch that we cant receive through the orderbook.go if we didn't update it

Reply:
I added a return above instead